pkg/client/client.go

```diff
@@ -618,85 +615,43 @@ func NewClient(ctrlRuntimeClient client.Client, secretName, namespace, regionID
-		if err != nil {
-			return nil, fmt.Errorf("failed to fetch  credentials file from Secret: %v", err)
-		}
-	}
-
-	return config, nil
-}
-
-// fetchCredentialsFileFromSecret fetch credentials from screct
-// alibabacloud accessKeyID accessKeySecret stsToken
-//roleArn, roleName ,roleSessionName ,roleSessionExpiration
-func fetchCredentialsFileFromSecret(secret *corev1.Secret, config *providers.Configuration) error {
-	//accessKeyID/accessKeySecret
-	if len(secret.Data[kubeAccessKeyID]) > 0 && len(secret.Data[kubeAccessKeySecret]) > 0 {
-		config.AccessKeyID = utils.ByteArray2String(secret.Data[kubeAccessKeyID])
-		config.AccessKeySecret = utils.ByteArray2String(secret.Data[kubeAccessKeySecret])
+func getCredentialFromSecret(ctrlRuntimeClient client.Client, secretName, namespace string, configManagedClient client.Client) (auth.Credential, error) {
+	if secretName == "" {
+		return nil, fmt.Errorf("secret name is empty")
 	}
-
-	//ststoken
-	if len(secret.Data[kubeAccessKeyStsToken]) > 0 {
-		config.AccessKeyStsToken = utils.ByteArray2String(secret.Data[kubeAccessKeyStsToken])
-	}
-
-	// roleArn ,roleSessionName ,roleSessionExpiration
-	if len(secret.Data[kubeRoleArn]) > 0 && len(secret.Data[kubeRoleSessionName]) > 0 && len(secret.Data[kubeRoleSessionExpiration]) > 0 {
-		if roleSessionExpiration, err := utils.String2IntPointer(string(secret.Data[kubeRoleSessionExpiration])); roleSessionExpiration != nil && err == nil {
-			config.RoleArn = string(secret.Data[kubeRoleArn])
-			config.RoleSessionName = string(secret.Data[kubeRoleSessionName])
-			config.RoleSessionExpiration = roleSessionExpiration
+	var secret corev1.Secret
+	if err := ctrlRuntimeClient.Get(context.Background(), client.ObjectKey{Namespace: namespace, Name: secretName}, &secret); err != nil {
+		if apimachineryerrors.IsNotFound(err) {
+			return nil, machineapiapierrors.InvalidMachineConfiguration("alibabacloud credentials secret %s/%s: %v not found", namespace, secretName, err)
 		}
+		return nil, err
 	}
+	return fetchCredentialsIniFromSecret(&secret)
+}
 
-	// roleName
-	if len(secret.Data[kubeRoleName]) > 0 {
-		config.RoleName = string(secret.Data[kubeRoleName])
-	}
-
-	return nil
-}
-
-// fetchCredentialsFileFromConfigMap fetch oleArn, roleName ,roleSessionName ,roleSessionExpiration from configmap
-// roleArn, roleName ,roleSessionName ,roleSessionExpiration
-func fetchCredentialsFileFromConfigMap(namespace string, configManagedClient client.Client, config *providers.Configuration) error {
-	cm := &corev1.ConfigMap{}
-	switch err := configManagedClient.Get(
-		context.Background(),
-		client.ObjectKey{Namespace: namespace, Name: kubeCloudConfigName},
-		cm,
-	); {
-	case apimachineryerrors.IsNotFound(err):
-		// no cloud config ConfigMap
-		return nil
-	case err != nil:
-		return fmt.Errorf("failed to get kube-cloud-config ConfigMap: %w", err)
+func fetchCredentialsIniFromSecret(secret *corev1.Secret) (auth.Credential, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
```

Comment:
Only the env parts of this need to be done under lock right? Should we move the lock call further down? Also worth a comment to explain why we are locking

Reply:
I have moved this down and commented.